kernel/src/actions/mod.rs

```diff
@@ -481,4 +555,66 @@ mod tests {
         )]));
         assert_eq!(schema, expected);
     }
+
+    #[test]
+    fn test_validate_protocol() {
+        let invalid_protocols = [
+            Protocol {
+                min_reader_version: 3,
+                min_writer_version: 7,
+                reader_features: None,
+                writer_features: Some(vec![]),
+            },
+            Protocol {
+                min_reader_version: 3,
+                min_writer_version: 7,
+                reader_features: Some(vec![]),
+                writer_features: None,
+            },
+            Protocol {
+                min_reader_version: 3,
+                min_writer_version: 7,
+                reader_features: None,
+                writer_features: None,
+            },
+        ];
+        for Protocol {
+            min_reader_version,
+            min_writer_version,
+            reader_features,
+            writer_features,
+        } in invalid_protocols
+        {
+            assert!(matches!(
+                Protocol::new(
+                    min_reader_version,
+                    min_writer_version,
+                    reader_features,
+                    writer_features
+                ),
+                Err(Error::InvalidProtocol(_)),
+            ));
+        }
+    }
+
+    #[test]
+    fn test_v2_checkpoint_unsupported() {
+        let protocol = Protocol::new(
+            3,
+            7,
+            Some(vec![ReaderFeatures::V2Checkpoint.to_string()]),
+            Some(vec![ReaderFeatures::V2Checkpoint.to_string()]),
```

Comment:
Since we anyway have a method now, maybe consider having it take:
```rust
reader_features: Option<impl` IntoIterator<Item = impl Into<String>>>,
writer_features: Option<impl` IntoIterator<Item = impl Into<String>>>,
```
... and then this code here simplifies to:
```suggestion
            Some([&ReaderFeatures::V2Checkpoint]),
            Some([&ReaderFeatures::V2Checkpoint]),
```
(again below)

Reply:
cool got it (and just added an `impl From<Reader/WriterFeatures> for String`)